🚀 RELEASE: v1.6.0 (#4816)
aiidateam/aiida-core: CHANGELOG.md

- The schema includes a number of new global/profile options, including: `transport.task_retry_initial_interval`, `transport.task_maximum_attempts`, `rmq.task_timeout` and `logging.aiopika_loglevel` ([#4583](https://github.com/aiidateam/aiida-core/pull/4583)).
- The `cache_config.yml` has now also been **deprecated** and merged into the `config.json`, as part of the profile options.
  This merge will be handled automatically, upon first load of the `config.json` using the new AiiDA version.

In-line with these changes, the `verdi config` command has been refactored into separate commands, including `verdi config list`, `verdi config set`, `verdi config unset` and `verdi config caching`.

See the [Configuring profile options](https://aiida.readthedocs.io/projects/aiida-core/en/v1.6.0/howto/installation.html#configuring-profile-options) and [Configuring caching](https://aiida.readthedocs.io/projects/aiida-core/en/v1.6.0/howto/run_codes.html#how-to-save-compute-time-with-caching) how-tos for more details.

### Command-line additions and improvements 👌

In addition to `verdi config`, numerous other new commands and options have been added to `verdi`:

> - **Deprecate** `verdi export` and `verdi import` commands and merge into `verdi archive` ([#4710](https://github.com/aiidateam/aiida-core/pull/4710))

Reviewer: ```suggestion
- **Deprecated** `verdi export` and `verdi import` commands (replaced by new `verdi archive`) ([#4710](https://github.com/aiidateam/aiida-core/pull/4710))
```

- Added `verdi group delete --delete-nodes`, to also delete the nodes in a group during its removal ([#4578](https://github.com/aiidateam/aiida-core/pull/4578)).

> - Improve `verdi group remove-nodes` command, to warn when requested nodes are not in the specified group ([#4728](https://github.com/aiidateam/aiida-core/pull/4728)).

Reviewer: ```suggestion
- Improved `verdi group remove-nodes` command to warn when requested nodes are not in the specified group ([#4728](https://github.com/aiidateam/aiida-core/pull/4728)).
```

- Added `exception` to the projection mapping of `verdi process list`, for example to use in debugging as: `verdi process list -S excepted -P ctime pk exception` ([#4786](https://github.com/aiidateam/aiida-core/pull/4786)).
- Added `verdi database summary` ([#4737](https://github.com/aiidateam/aiida-core/pull/4737)):
  This prints a summary of the count of each entity and (optionally) the list of unique identifiers for some entities.
- Improved `verdi process play` performance, by only querying for active processes with the `--all` flag ([#4671](https://github.com/aiidateam/aiida-core/pull/4671))
- Added the `verdi database version` command ([#4613](https://github.com/aiidateam/aiida-core/pull/4613)):
  This shows the schema generation and version of the database of the given profile, useful mostly for developers when debugging.
- Improved `verdi node delete` performance ([#4575](https://github.com/aiidateam/aiida-core/pull/4575)):

>   Theed `verdi quicksetup --non-interactive`, to ensure it does not include any user prompts ([#4573](https://github.com/aiidateam/aiida-core/pull/4573))

Reviewer: "Theed"? Is this a word I don't know or is this a cropping error?

Author: fixed regression

> - Improved `verdi node delete` performance ([#4575](https://github.com/aiidateam/aiida-core/pull/4575)):
>   Theed `verdi quicksetup --non-interactive`, to ensure it does not include any user prompts ([#4573](https://github.com/aiidateam/aiida-core/pull/4573))

Author: ```suggestion
- Improved `verdi node delete` performance ([#4575](https://github.com/aiidateam/aiida-core/pull/4575)):
  The logic has been re-written to greatly reduce the time to delete large amounts of nodes.
- Fixed `verdi quicksetup --non-interactive`, to ensure it does not include any user prompts ([#4573](https://github.com/aiidateam/aiida-core/pull/4573))
```

### New REST API Query endpoint ✨

The `/querybuilder` endpoint is the first POST method available for AiiDA's RESTful API ([#4337](https://github.com/aiidateam/aiida-core/pull/4337))

The POST endpoint returns what the QueryBuilder would return, when providing it with a proper `queryhelp` dictionary ([see the documentation here](https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/database.html#the-queryhelp)).
Furthermore, it returns the entities/results in the "standard" REST API format - with the exception of `link_type` and `link_label` keys for links (these particular keys are still present as `type` and `label`, respectively).

For security, POST methods can be toggled on/off with the `verdi restapi --posting/--no-posting` options (it is on by default).
Although note that this option is not yet strictly public, since its naming may be changed in the future!

See [AiiDA REST API documentation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/reference/rest_api.html) for more details.

### Additional Changes

> - Fix direct scheduler, in combination with `SshTransport`, hanging on submit command ([#4735](https://github.com/aiidateam/aiida-core/pull/4735)).

Reviewer: ```suggestion
- Fixed the direct scheduler which, in combination with `SshTransport`, was hanging on submit command ([#4735](https://github.com/aiidateam/aiida-core/pull/4735)).
```

  In the ssh transport, to emulate 'chdir', the current directory is now kept in memory, and every command prepended with `cd FOLDER_NAME && ACTUALCOMMAND`.

> - In `aiida.tools.ipython.ipython_magics`, `load_ipython_extension` is deprecated in favour of `register_ipython_extension` ([#4548](https://github.com/aiidateam/aiida-core/pull/4548)).

Reviewer: ```suggestion
- In `aiida.tools.ipython.ipython_magics`, `load_ipython_extension` has been **deprecated** in favour of `register_ipython_extension` ([#4548](https://github.com/aiidateam/aiida-core/pull/4548)).
```

> - Refactor `.ci/` folder, tests more portable and easier to understand ([#4565](https://github.com/aiidateam/aiida-core/pull/4565))

Reviewer: ```suggestion
- Refactored `.ci/` folder to make tests more portable and easier to understand ([#4565](https://github.com/aiidateam/aiida-core/pull/4565))
```

# Changelog

## v1.6.0 - 2021-03-15

[full changelog](https://github.com/aiidateam/aiida-core/compare/v1.5.2...v1.6.0) | [GitHub contributors page for this release](https://github.com/aiidateam/aiida-core/graphs/contributors?from=2020-12-07&to=2021-03-15&type=c)

As well as introducing a number of improvements and new features listed below,
this release marks the "under-the-hood" migration from the `tornado` package to the Python built-in module `asyncio`,

> for handling asynchronous processing within the AiiDA engine.

Author: done

This removes a number of blocking dependency version clashes with other tools,
in particular with the newest Jupyter shell and notebook environments.
The migration does not present any backward incompatible changes to AiiDA's public API.
A substantial effort has been made to test and debug the new implementation, and ensure it performs at least equivalent to the previous code (or improves it!), but please let us know if you uncover any additional issues.

This release also drops support for Python 3.6 (testing is carried out against `3.7`, `3.8` and `3.9`).

NOTE: `v1.6` is tentatively intended to be the final minor `v1.x` release before `v2.0.0`;

> introducing a new (back-incompatible) file repository and removing all deprecated code.

Author: reworded

### New calculation features ✨

Add the `additional_retrieve_list` metadata option ([#4437](https://github.com/aiidateam/aiida-core/pull/4437)).
This new option allows one to specify additional files to be retrieved on a per-instance basis, in addition to the files that are already defined by the plugin to be retrieved.

> A **new namespace `stash`** is added to the `metadata.options` input

Author: fair

> A **new namespace `stash`** is added to the `metadata.options` input

Author: done

- **Deprecate** `verdi export` and `verdi import` commands and merge into `verdi archive` ([#4710](https://github.com/aiidateam/aiida-core/pull/4710))
- Add `verdi group delete --delete-nodes`, to also delete the nodes in a group during its removal ([#4578](https://github.com/aiidateam/aiida-core/pull/4578)).
- Improve `verdi group remove-nodes` command, to warn when requested nodes are not in the specified group ([#4728](https://github.com/aiidateam/aiida-core/pull/4728)).
- Add `verdi database summary` ([#4737](https://github.com/aiidateam/aiida-core/pull/4737)):
  This prints a summary of the count of each entity and (optionally) the list of unique identifiers for some entities.
- Improve `verdi process play` performance, by only querying for active processes with the `--all` flag ([#4671](https://github.com/aiidateam/aiida-core/pull/4671))
- Add the `verdi database version` command ([#4613](https://github.com/aiidateam/aiida-core/pull/4613)):
  This shows the schema generation and version of the database of the given profile, useful mostly for developers when debugging.
- Improve `verdi node delete` performance ([#4575](https://github.com/aiidateam/aiida-core/pull/4575)):

>   The logic has been re-written to greatly reduce the time to delete large amounts of nodes.

Author: Hmm, I'll consider, but the current order was IMO order of importance, e.g. `verdi database version` has very niche application, whereas I think `verdi database summary` will be helpful for a lot more people

>   The logic has been re-written to greatly reduce the time to delete large amounts of nodes.

Reviewer: Ah, I see. Well, you can also see how it is once you add some of the extra commits in the list. I personally get a better global idea of the changes if I can see the related topics together; if I see them listed too separately my brain has to be switching topics and is harder to keep track. I agree an importance criteria could be used then to arrange the list internally of each topic or the topics themselves. But yeah, maybe try it out and decide what seems better.

- Fix `verdi quicksetup --non-interactive`, to ensure it does not include any user prompts ([#4573](https://github.com/aiidateam/aiida-core/pull/4573))
- Fix `verdi --version` when used in editable mode ([#4576](https://github.com/aiidateam/aiida-core/pull/4576))

### API additions and improvements 👌

Base `Node` `__eq__` and `__hash__` methods are now based on a node's UUID ([#4753](https://github.com/aiidateam/aiida-core/pull/4753)).

> For example, loading the same node twice will always resolve as equivalent: `load_node(1) == load_node(1)`, and nodes can be correctly added to sets: `len({load_node(1), load_node(1)}) == 1`.

Author: hence why I said the commit message should have been better 😆 

> We just set the `__hash__` for the nodes that are not mocking python intrinsics

Oh thats a bit crap: Note, from ticket [1549](https://bugs.python.org/issue1549):

> This was done intentionally -- if you define a comparison without defining a hash, the default hash will not match your comparison, and your objects will misbehave when used as dictionary keys. (Guido van Rossum)

> For example, loading the same node twice will always resolve as equivalent: `load_node(1) == load_node(1)`, and nodes can be correctly added to sets: `len({load_node(1), load_node(1)}) == 1`.

Reviewer: Mmm maybe. Anyways, just for context: we were not trying to go against the intentionality of the feature, this was already behaving like that before because of the special definition of `__eq__` in the python type mocking node types. In the modification we did for the  `Node` base class we set the `__hash__` to behave the same as the `__eq__` for the fallback. We didn't do anything to the subclass because of the [unresolved discussion](https://github.com/aiidateam/aiida-core/issues/1917) of how that should behave so we just let it as it was.

> For example, loading the same node twice will always resolve as equivalent: `load_node(1) == load_node(1)`, and nodes can be correctly added to sets: `len({load_node(1), load_node(1)}) == 1`.

Author: done

Type annotations have been added to all inputs/outputs of functions and methods in `aiida.engine` ([#4669](https://github.com/aiidateam/aiida-core/pull/4669)) and `aiida/orm/nodes/processes` ([#4772](https://github.com/aiidateam/aiida-core/pull/4772)).
As outlined in [PEP 484](https://www.python.org/dev/peps/pep-0484/), this improves static code analysis and, for example, allows for better auto-completion and type checking in many code editors.

### New REST API Query endpoint ✨

The `/querybuilder` endpoint is the first POST method available for AiiDA's RESTful API ([#4337](https://github.com/aiidateam/aiida-core/pull/4337))

The POST endpoint returns what the QueryBuilder would return, when providing it with a proper queryhelp dictionary ([see the documentation here](https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/database.html#the-queryhelp)).
Furthermore, it returns the entities/results in the "standard" REST API format - with the exception of `link_type` and `link_label` keys for links (these particular keys are still present as `type` and
`label`, respectively).

> For security, POST methods can be toggled on/off with the `verdi restapi --posting/--no-posting` options.

Author: 👍

> For security, POST methods can be toggled on/off with the `verdi restapi --posting/--no-posting` options.

Author: done
